File: 백준/Gold/2096. 내려가기/내려가기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        
        int[] minDp = new int[3];
        int[] maxDp = new int[3];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int[] num = new int[3];
            for (int j = 0; j < 3; j++) {
                num[j] = Integer.parseInt(st.nextToken());
            }
            
            int[] minTmp = minDp.clone();
            int[] maxTmp = maxDp.clone();
            
            minDp[0] = Math.min(minTmp[0], minTmp[1]) + num[0];
            maxDp[0] = Math.max(maxTmp[0], maxTmp[1]) + num[0];
            
            minDp[1] = Math.min(Math.min(minTmp[0], minTmp[1]), minTmp[2]) + num[1];
            maxDp[1] = Math.max(Math.max(maxTmp[0], maxTmp[1]), maxTmp[2]) + num[1];
            
            minDp[2] = Math.min(minTmp[1], minTmp[2]) + num[2];
            maxDp[2] = Math.max(maxTmp[1], maxTmp[2]) + num[2];
        }
        
        int maxResult = Math.max(Math.max(maxDp[0], maxDp[1]), maxDp[2]);
        int minResult = Math.min(Math.min(minDp[0], minDp[1]), minDp[2]);
        
        System.out.println(maxResult + " " + minResult);
    }
}

File: 백준/Gold/2096. 내려가기/내려가기.py
# 내려가기
import sys
input = sys.stdin.readline

n = int(input())
min_dp = [0, 0, 0]
max_dp = [0, 0, 0]

for i in range(n):
    num = list(map(int,input().split()))
    min_tmp = min_dp.copy()
    max_tmp = max_dp.copy()
    
    for j in range(3):
        if j == 0:
            min_dp[j] = min(min_tmp[0], min_tmp[1]) + num[j]
            max_dp[j] = max(max_tmp[0], max_tmp[1]) + num[j]
        elif j == 1:
            min_dp[j] = min(min_tmp[0], min_tmp[1], min_tmp[2]) + num[j]
            max_dp[j] = max(max_tmp[0], max_tmp[1], max_tmp[2]) + num[j]
        else:
            min_dp[j] = min(min_tmp[1], min_tmp[2]) + num[j]
            max_dp[j] = max(max_tmp[1], max_tmp[2]) + num[j]

print(max(max_dp), min(min_dp))
